**Ejercicio 1:** Selecciona una API pública y extrae datos utilizando Python.

Tengo como afición la astronomía y me ha parecido realizar este ejercicio con algo relacionado al tema.

Selecciono la API de la NASA, del Servicio de Detección de objetos Cercanos a la Tierra (NeoWs) que se encuentra dentro de la página https://api.nasa.gov/ para ver si puedo obtener algunos datos relacionados con objetos entre los días 17 y 18 de octubre de 2024.


In [1]:
import requests
import pandas as pd

# Hago la solicitud a  la API de la NASA 

URL = "https://api.nasa.gov/neo/rest/v1/feed?start_date=2024-10-17&end_date=2024-10-18&api_key=DEMO_KEY"
response = requests.get(URL)

if response.status_code == 200:
    data = response.json()
    print("Solicitud Exitosa")
    print("Data:", data)
else:
    print("Error en la solicitud, detalles:", response.text)

# Reviso como están estructurados los datos
print(data.keys())



Solicitud Exitosa
Data: {'links': {'next': 'http://api.nasa.gov/neo/rest/v1/feed?start_date=2024-10-18&end_date=2024-10-19&detailed=false&api_key=DEMO_KEY', 'previous': 'http://api.nasa.gov/neo/rest/v1/feed?start_date=2024-10-16&end_date=2024-10-17&detailed=false&api_key=DEMO_KEY', 'self': 'http://api.nasa.gov/neo/rest/v1/feed?start_date=2024-10-17&end_date=2024-10-18&detailed=false&api_key=DEMO_KEY'}, 'element_count': 40, 'near_earth_objects': {'2024-10-18': [{'links': {'self': 'http://api.nasa.gov/neo/rest/v1/neo/2483547?api_key=DEMO_KEY'}, 'id': '2483547', 'neo_reference_id': '2483547', 'name': '483547 (2003 WM25)', 'nasa_jpl_url': 'https://ssd.jpl.nasa.gov/tools/sbdb_lookup.html#/?sstr=2483547', 'absolute_magnitude_h': 17.65, 'estimated_diameter': {'kilometers': {'estimated_diameter_min': 0.7844314124, 'estimated_diameter_max': 1.7540419619}, 'meters': {'estimated_diameter_min': 784.4314124479, 'estimated_diameter_max': 1754.0419619197}, 'miles': {'estimated_diameter_min': 0.487422

In [2]:
# Dentro del campo near_earth_objects es donde puedo encontrar los datos importantes(contiene los asteroides categorizados)

near_earth_objects = data['near_earth_objects']

# Extraigo la información de cada objeto cercano de las fechas solicitadas

all_neo_data = []
for date in near_earth_objects:
    for neo in near_earth_objects[date]:
        neo_data = {
            'id': neo['id'],
            'nombre': neo['name'],
            'nasa_jpl_url': neo['nasa_jpl_url'],
            'magnitud_absoluta_h': neo['absolute_magnitude_h'],
            'diametro_min_estimado_km': neo['estimated_diameter']['kilometers']['estimated_diameter_min'],
            'diametro_max_estimado_km': neo['estimated_diameter']['kilometers']['estimated_diameter_max'],
            'potencialmente_peligroso': neo['is_potentially_hazardous_asteroid'],
            'fecha_aproximacion_cercana': neo['close_approach_data'][0]['close_approach_date'],
            'velocidad_relativa_km_per_seg': neo['close_approach_data'][0]['relative_velocity']['kilometers_per_second'],
            'distancia_fallo_km': neo['close_approach_data'][0]['miss_distance']['kilometers'],
            'cuerpo_que_orbita': neo['close_approach_data'][0]['orbiting_body']
    }
    all_neo_data.append(neo_data)

# Convierto los datos en un Dataframe e imprimo el resultado

df_neo = pd.DataFrame(all_neo_data)

print(df_neo)




         id       nombre                                       nasa_jpl_url  \
0  54494098  (2024 TB22)  https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...   
1  54491553  (2024 TC15)  https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...   

   magnitud_absoluta_h  diametro_min_estimado_km  diametro_max_estimado_km  \
0               19.323                  0.363040                  0.811783   
1               26.862                  0.011276                  0.025214   

   potencialmente_peligroso fecha_aproximacion_cercana  \
0                      True                 2024-10-18   
1                     False                 2024-10-17   

  velocidad_relativa_km_per_seg  distancia_fallo_km cuerpo_que_orbita  
0                 22.0048475276  11433124.266985772             Earth  
1                  9.2205371483   1835002.284844349             Earth  


Con la información del dataset puedo destacar que el objeto ***2024 TB22*** se considera potencialmente peligroso, explico los valores obtenidos de este objeto:

- **Magnitud absoluta:** 19.32 (la brillo del objeto, cuando menor el número más brillantes y por lo tanto más grande, ejemplo: el planeta Venus tiene una magnitud de -4,38, una estrella como Polaris, la estrella polar es de +1.97 )

- **Diámetro estimado:** entre 363 y 812 mts. De impactar en la tierra podría causar un daño regional significativo.

- **Potencialmente peligroso:** True, la NASA lo considera potencialmente peligroso.

- **Fecha de aproximación cercana:** 18 de octubre de 2024

- **Velocidad relativa:** 22 km/s velocidad extremadamente alta en relación con la tierra, en caso de impacto la energía sería considerable.

- **Distancia de fallo:** 11 millones de km. aunque esté clasificado como potencialmente peligroso, pasará a una distancia segura (más lejos que la distancia entre la tierra y la luna)

- **Trayectoria:** tiene una trayectoria que lo llevará cerca de la tierra.